go/vt/vtctld/redirection.go

```diff
@@ -0,0 +1,79 @@
+/*
+Copyright 2020 The Vitess Authors.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package vtctld
+
+import (
+	"bytes"
+	"fmt"
+	"io/ioutil"
+	"net/http"
+	"net/http/httputil"
+	"net/url"
+	"strings"
+
+	"vitess.io/vitess/go/vt/log"
+	"vitess.io/vitess/go/vt/topo"
+	"vitess.io/vitess/go/vt/topo/topoproto"
+)
+
+func initVTTabletRedirection(ts *topo.Server) {
+	http.HandleFunc("/vttablet/", func(w http.ResponseWriter, r *http.Request) {
+		splits := strings.SplitN(r.URL.Path, "/", 4)
+		if len(splits) < 4 {
+			log.Errorf("Invalid URL: %v", r.URL)
+			http.NotFound(w, r)
+			return
+		}
+		tabletID := splits[2]
+		tabletAlias, err := topoproto.ParseTabletAlias(tabletID)
+		if err != nil {
+			log.Errorf("Error parsting tablet alias %v: %v", tabletID, err)
+			http.NotFound(w, r)
+			return
+		}
+		tablet, err := ts.GetTablet(r.Context(), tabletAlias)
+		if err != nil {
+			log.Errorf("Error fetching tablet %v: %v", splits[2], err)
+			http.NotFound(w, r)
+			return
+		}
+		redirect := fmt.Sprintf("http://%s:%d", tablet.Hostname, tablet.PortMap["vt"])
+		u, err := url.Parse(redirect)
+		if err != nil {
+			log.Errorf("Error parsing url %s: %v", redirect, err)
+			http.NotFound(w, r)
+			return
+		}
+		rp := httputil.NewSingleHostReverseProxy(u)
+		prefixPath := fmt.Sprintf("/vttablet/%s/", tabletID)
+		rp.ModifyResponse = func(r *http.Response) error {
+			b, _ := ioutil.ReadAll(r.Body)
+			b = bytes.ReplaceAll(b, []byte(`href="/`), []byte(fmt.Sprintf(`href="%s`, prefixPath)))
+			b = bytes.ReplaceAll(b, []byte(`href=/`), []byte(fmt.Sprintf(`href=%s`, prefixPath)))
+			r.Body = ioutil.NopCloser(bytes.NewBuffer(b))
+			r.Header["Content-Length"] = []string{fmt.Sprint(len(b))}
+			// Don't forget redirects
+			loc := r.Header["Location"]
+			for i, v := range loc {
+				loc[i] = strings.Replace(v, "/", prefixPath, 1)
+			}
+			return nil
+		}
+		r.URL.Path = "/" + splits[3]
```

Comment:
We should do this in a `Director` function since `ReverseProxy` requires one anyway. I'm guessing you haven't tested this rewrite yet since it should panic in its current form...

Also, for what it's worth, `ReverseProxy` makes a deep copy of the `Request` before letting you modify it.

Reply:
What use case would cause this to panic? It worked fine when I tested this.

Reply:
Oh you're right. I was mistaken. I was thinking about a case when I previously had tried to set `Director = nil` explicitly since I wanted a no-op Director.